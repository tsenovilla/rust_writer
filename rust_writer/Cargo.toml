[package]
name = "rust_writer"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
prettyplease = { workspace = true }
proc-macro2 = { workspace = true }
regex = { workspace = true }
rust_writer_proc = "1.0.3"
rustilities = { workspace = true, features = ["parsing"] }
syn = { workspace = true, features = ["full", "extra-traits", "visit", "visit-mut"] }
thiserror = { workspace = true }

[dev-dependencies]
tempfile = "3.16.0"
test_builder = { workspace = true }
quote = { workspace = true }
