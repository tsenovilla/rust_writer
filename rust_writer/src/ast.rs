// SPDX-License-Identifier: GPL-3.0

pub mod finder;
pub mod implementors;
pub mod macros;
pub mod mutator;
