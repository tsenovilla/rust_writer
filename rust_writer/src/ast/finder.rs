// SPDX-License-Identifier: GPL-3.0

//! This module contains the [`Finder`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/struct.Finder.html)
//! struct, which is used to search for specific items in an AST in a very targeted way.
//! The `Finder` struct is completely generic, and thanks to the
//! [`ToFind`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/trait.ToFind.html) trait and the
//! [implementors](https://docs.rs/rust_writer/latest/rust_writer/ast/implementors/index.html) it
//! may be customized to assert if an element is contained inside an AST.
//!
//! # Example
//! ```rust
//! use test_builder::TestBuilder;
//! use rust_writer::ast::{
//!     finder::{Finder, ToFind},
//!     implementors::ItemToTrait
//! };
//! use syn::{
//!     visit::Visit,
//!     parse_quote,
//!     File
//! };
//!
//! TestBuilder::default()
//!     .with_trait_ast()
//!     .execute(|builder| {
//!         let ast = builder.get_ref_ast_file("trait.rs").expect("This exists; qed;");
//!
//!         // Define the ItemFinder implementor. This implementor specifies:
//!         // 1. The trait we want to search for, e.g., a trait named `MyTrait`.
//!         // 2. The specific item(s) we expect to find within that trait.
//!         let item_to_trait: ItemToTrait =
//!             ("MyTrait", parse_quote! { type Type1: From<String>; }).into();
//!
//!         // Create the Finder and load this particular implementor
//!         let mut finder = Finder::default().to_find(&item_to_trait);
//!
//!         // Search the AST for the expected item
//!         assert!(finder.find(&ast));
//!     });
//! ```

#[cfg(test)]
mod tests;

use std::fmt::Debug;
use syn::{visit::Visit, File};

/// A placeholder finder which does not perform any specific search.
/// This is used as the default for a [`Finder`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/struct.Finder.html).
#[derive(Debug)]
pub struct EmptyFinder;

/// The `Finder` struct is used to assert if a concrete element belongs to an AST.
/// The `Finder` struct is totally generic, so it can be thought of a game console: its purpose is
/// to determine if an element is contained inside an AST but it doesn't know how to do it until an
/// [implementor](https://docs.rs/rust_writer/latest/rust_writer/ast/implementors/index.html)
/// is loaded. This implementor will tell `Finder` how it should behave.
///
/// A new `Finder` can be easily created using `Finder::default()`, which returns a
/// `Finder<'_, EmptyFinder, 1>`. This `Finder` is useless up to this point, but thanks to the
/// [`ToFind`](https://docs.rs/rust_writer/latest/rust_writer/ast/finder/trait.ToFind.html)
/// trait, an implementor can be seamlessly loaded to the `Finder`, releasing its whole power.
///
/// Once configured, the [`find`](#method.find) method can be called to look inside the AST.
///
/// ```rust
/// use test_builder::TestBuilder;
/// use rust_writer::ast::{
///     finder::{Finder, ToFind},
///     implementors::ItemToTrait
/// };
/// use syn::{
///     visit::Visit,
///     parse_quote,
///     File
/// };
///
/// TestBuilder::default()
///     .with_trait_ast()
///     .execute(|builder| {
///         let ast = builder.get_ref_ast_file("trait.rs").expect("This exists; qed;");
///
///         // Define the ItemFinder implementor. This implementor specifies:
///         // 1. The trait we want to search for, e.g., a trait named `MyTrait`.
///         // 2. The specific item(s) we expect to find within that trait.
///         let item_to_trait: ItemToTrait =
///             ("MyTrait", parse_quote! { type Type1: From<String>; }).into();
///
///         // Create the Finder and load this particular implementor
///         let mut finder = Finder::default().to_find(&item_to_trait);
///
///         // Search the AST for the expected item
///         assert!(finder.find(&ast));
///     });
/// ```
#[derive(Debug, Clone)]
pub struct Finder<'a, T: Debug, const N: usize> {
	/// A `Finder` can act in different parts of the AST at the same time. This array keeps track
	/// of which of these searchs succeeded.  
	pub found: [bool; N],
	/// Placeholder to load an implementor.
	pub finder: &'a T,
}

/// This trait is used to create a new `Finder` variable.
/// It is typically implemented for `Finder<'_, EmptyFinder, 1>`, enabling the flow:
/// 1. Create a `Finder<'_, EmptyFinder, 1>` using `Finder::default()`.
/// 2. Configure it with the desired implementor using the `to_find` method.
pub trait ToFind<'a, T: Debug, const N: usize> {
	fn to_find(self, finder: &'a T) -> Finder<'a, T, N>;
}

impl Default for Finder<'_, EmptyFinder, 1> {
	fn default() -> Self {
		Self { found: [false], finder: &EmptyFinder }
	}
}

impl<'a, T, const N: usize> Finder<'a, T, N>
where
	T: Debug,
	Finder<'a, T, N>: Visit<'a>,
{
	/// Apply all the searches defined by the implementor.
	pub fn find(&mut self, ast: &'a File) -> bool {
		self.visit_file(ast);
		self.found.iter().all(|&x| x)
	}
}

impl<T, const N: usize> Finder<'_, T, N>
where
	T: Debug,
{
	/// Reset the `found` array to all `false` values.
	pub fn reset(&mut self) {
		self.found = [false; N];
	}
}
