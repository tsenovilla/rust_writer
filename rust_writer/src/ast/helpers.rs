// SPDX-License-Identifier: GPL-3.0

#[cfg(test)]
mod tests;

use syn::{Attribute, ImplItem, Item, TraitItem};

pub(crate) trait AttrsMut {
	fn attrs_mut(&mut self) -> Option<&mut Vec<Attribute>>;
}

pub(crate) fn item_without_docs<T: AttrsMut + Clone>(item: &T) -> T {
	let mut output = item.clone();
	if let Some(attrs) = output.attrs_mut() {
		attrs.retain(|attr| !attr.path().is_ident("doc"));
	}
	output
}

impl AttrsMut for Item {
	fn attrs_mut(&mut self) -> Option<&mut Vec<Attribute>> {
		match self {
			Item::Const(item) => Some(&mut item.attrs),
			Item::Enum(item) => Some(&mut item.attrs),
			Item::ExternCrate(item) => Some(&mut item.attrs),
			Item::Fn(item) => Some(&mut item.attrs),
			Item::ForeignMod(item) => Some(&mut item.attrs),
			Item::Impl(item) => Some(&mut item.attrs),
			Item::Macro(item) => Some(&mut item.attrs),
			Item::Mod(item) => Some(&mut item.attrs),
			Item::Static(item) => Some(&mut item.attrs),
			Item::Struct(item) => Some(&mut item.attrs),
			Item::Trait(item) => Some(&mut item.attrs),
			Item::TraitAlias(item) => Some(&mut item.attrs),
			Item::Type(item) => Some(&mut item.attrs),
			Item::Union(item) => Some(&mut item.attrs),
			Item::Use(item) => Some(&mut item.attrs),
			_ => None,
		}
	}
}

impl AttrsMut for ImplItem {
	fn attrs_mut(&mut self) -> Option<&mut Vec<Attribute>> {
		match self {
			ImplItem::Const(item) => Some(&mut item.attrs),
			ImplItem::Fn(item) => Some(&mut item.attrs),
			ImplItem::Type(item) => Some(&mut item.attrs),
			ImplItem::Macro(item) => Some(&mut item.attrs),
			_ => None,
		}
	}
}

impl AttrsMut for TraitItem {
	fn attrs_mut(&mut self) -> Option<&mut Vec<Attribute>> {
		match self {
			TraitItem::Const(item) => Some(&mut item.attrs),
			TraitItem::Fn(item) => Some(&mut item.attrs),
			TraitItem::Type(item) => Some(&mut item.attrs),
			TraitItem::Macro(item) => Some(&mut item.attrs),
			_ => None,
		}
	}
}
