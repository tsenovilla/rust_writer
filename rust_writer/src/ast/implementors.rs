// SPDX-License-Identifier: GPL-3.0

mod item_to_impl;
mod item_to_mod;
mod item_to_trait;
mod token_stream_to_macro;

pub use item_to_impl::ItemToImpl;
pub use item_to_mod::ItemToMod;
pub use item_to_trait::ItemToTrait;
pub use token_stream_to_macro::TokenStreamToMacro;
