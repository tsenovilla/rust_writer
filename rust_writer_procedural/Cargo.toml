[package]
name = "rust_writer_procedural"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["full", "extra-traits", "parsing"] }

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.103"
rust_writer = { workspace = true }
test_builder = { workspace = true }
