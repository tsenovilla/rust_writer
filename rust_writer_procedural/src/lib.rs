// SPDX-License-Identifier: GPL-3.0

mod finder;
pub(crate) mod helpers;
mod mutator;
pub(crate) mod parse;

use proc_macro::TokenStream;

/// # Description
///
/// The `#[mutator]` macro is used to define a new implementor which combines other implementors,
/// capable of mutating different parts of an AST with just one instruction, with the same level of
/// precision as if we had used each implementor separately.
///
/// The syntax to use the `#[mutator]` macro is pretty simple:
///
/// ```rust
/// use rust_writer::ast::{mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::visit_mut::VisitMut;
///
/// // The new implementor behaves as two `ItemToTrait` implementors.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>)]
/// struct NewImplementor;
/// ```
///
/// From the snippet above we can already assert that the `#[mutator]` macro needs that the traits
/// [`ToMutate`](https://docs.rs/rust_writer/latest/rust_writer/ast/mutator/trait.ToMutate.html)
/// and [`VisitMut`](https://docs.rs/syn/latest/syn/visit_mut/trait.VisitMut.html) need to be
/// brought into scope.
///
/// It's also possible to combine an implementor from the predefined set with a local implementor,
/// just by adding the keyword `local` before the local implementor.
///
/// ```rust
/// use rust_writer::ast::{mutator, local_mutator, implementors::ItemToTrait, mutator::ToMutate};
/// use syn::{visit_mut::VisitMut, File};
///
/// #[local_mutator]
/// #[derive(Debug, Clone)]
/// struct LocalImplementor<T: std::fmt::Debug + Clone>{
///   mutated: [bool; 4],
///   something: T
/// }
///
/// impl<T> VisitMut for LocalImplementor<T> where T: std::fmt::Debug + Clone{
///   fn visit_file_mut(&mut self, _file: &mut File){
///     // Dummy implementation
///     self.mutated = [true; 4];
///   }
/// }
///
/// // The new implementor behaves as two `ItemToTrait` implementors + our local implementor.
/// #[mutator(ItemToTrait<'a>, ItemToTrait<'a>, local = LocalImplementor<T: std::fmt::Debug + Clone>)]
/// struct NewImplementor;
/// ```
///
/// When the `#[mutator]` macro is called, the struct is expanded an includes a new field for each
/// implementor named as the implementor type in lowercase. If an implementor is included several
/// times in the same macro invocation, a suffix is added to the subsequent fields, such as
/// _1, _2, _3 and so on. For example, the snippet above struct will look like this after
/// the expansion:
///
/// ```ignore
/// struct NewImplementor<'a, T: std::fmt::Debug + Clone>{
///   itemtotrait: ItemToTrait<'a>,
///   itemtotrait_1: ItemToTrait<'a>,
///   localimplementor: LocalImplementor<T>
/// }
/// ```
///
/// Awesome, these snippets also show that if an implementor needs a generic, including it in the
/// macro invocation is enough. Particularly, the generics will be added to the `NewImplementor`
/// struct, so they'll be available inside it. It's also possible to place.
///
/// # The `#[impl_from]` attribute
///
/// # Compatibility with [`#[finder]`](https://docs.rs/rust_writer/latest/rust_writer/ast/attr.finder.html)
#[proc_macro_attribute]
pub fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	mutator::mutator(attrs, item)
}

#[proc_macro_attribute]
pub fn local_mutator(_: TokenStream, item: TokenStream) -> TokenStream {
	mutator::local_mutator(item)
}

#[proc_macro_attribute]
pub fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::finder(attrs, item)
}

#[proc_macro_attribute]
pub fn local_finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::local_finder(attrs, item)
}

#[doc(hidden)]
#[proc_macro_attribute]
pub fn already_expanded(_: TokenStream, item: TokenStream) -> TokenStream {
	item
}
