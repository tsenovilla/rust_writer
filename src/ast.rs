// SPDX-License-Identifier: GPL-3.0

pub mod finder;
pub mod implementors;
pub mod mutator;
