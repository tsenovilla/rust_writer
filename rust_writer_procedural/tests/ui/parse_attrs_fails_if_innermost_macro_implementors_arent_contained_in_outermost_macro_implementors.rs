// SPDX-License-Identifier: GPL-3.0

use rust_writer::ast::{
	finder::ToFind,
	implementors::{ItemToImpl, ItemToTrait},
};
use rust_writer_procedural::{finder, impl_mutator, mutator};
use syn::{visit::Visit, visit_mut::VisitMut};

#[impl_mutator]
#[derive(Debug)]
struct SomeImplementor {
	mutated: [bool; 1],
}

impl VisitMut for SomeImplementor {
	fn visit_item_trait_mut(&mut self, _item_trait: &mut syn::ItemTrait) {}
}

#[finder(ItemToImpl<'a>, ItemToTrait<'a>)]
#[mutator(ItemToTrait<'a>, local = SomeImplementor, ItemToImpl<'a>)]
#[impl_from]
struct SomeStruct;

fn main() {}
