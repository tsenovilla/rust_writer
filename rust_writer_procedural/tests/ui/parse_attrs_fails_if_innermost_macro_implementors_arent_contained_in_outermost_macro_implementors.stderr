error: #[impl_from] is only allowed in structs annotated with #[mutator]/#[finder] at most once or structs annotated with #[mutator] and #[finder] such that the outermost macro implementors set contains the innermost macro implementors set.
  --> tests/ui/parse_attrs_fails_if_innermost_macro_implementors_arent_contained_in_outermost_macro_implementors.rs:23:8
   |
23 | struct SomeStruct;
   |        ^^^^^^^^^^

error[E0412]: cannot find type `SomeStruct` in this scope
  --> tests/ui/parse_attrs_fails_if_innermost_macro_implementors_arent_contained_in_outermost_macro_implementors.rs:23:8
   |
23 | struct SomeStruct;
   |        ^^^^^^^^^^ not found in this scope
